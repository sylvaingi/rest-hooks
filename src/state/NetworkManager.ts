import { memoize } from 'lodash';
import { FetchAction, ReceiveAction, MiddlewareAPI, Manager } from '../types';

export const RIC: (cb: (...args: any[]) => void, options: any) => void =
  typeof (global as any).requestIdleCallback === 'function'
    ? (global as any).requestIdleCallback
    : cb => global.setTimeout(cb, 0);

/** Handles all async network dispatches
 *
 * Dedupes concurrent requests by keeping track of all fetches in flight
 * and returning existing promises for requests already in flight.
 *
 * Interfaces with store via a redux-compatible middleware.
 */
export default class NetworkManager implements Manager {
  protected fetched: { [k: string]: Promise<any> } = {};
  protected resolvers: { [k: string]: (value?: any) => void } = {};
  protected rejectors: { [k: string]: (value?: any) => void } = {};
  readonly dataExpiryLength: number;
  readonly errorExpiryLength: number;
  constructor(
    dataExpiryLength: number = 60000,
    errorExpiryLength: number = 1000,
  ) {
    this.dataExpiryLength = dataExpiryLength;
    this.errorExpiryLength = errorExpiryLength;
  }

  /** Ensures all promises are completed by rejecting remaining. */
  cleanup() {
    for (const k in this.rejectors) {
      this.rejectors[k](new Error('Cleaning up Network Manager'));
    }
  }

  /** Clear promise state for a given url */
  protected clear(url: string) {
    delete this.resolvers[url];
    delete this.rejectors[url];
    delete this.fetched[url];
  }

  /** Called when middleware intercepts 'rest-hooks/fetch' action.
   *
   * Will then start a promise for a url and potentially start the network
   * fetch.
   *
   * Uses throttle only when instructed by action meta. This is valuable
   * for ensures mutation requests always go through.
   */
  protected handleFetch(action: FetchAction, dispatch: React.Dispatch<any>) {
    const fetch = action.payload;
    const {
      schema,
      url,
      responseType,
      throttle,
      resolve,
      reject,
      options = {},
    } = action.meta;
    const {
      dataExpiryLength = this.dataExpiryLength,
      errorExpiryLength = this.errorExpiryLength,
    } = options;

    const deferedFetch = () =>
      fetch()
        .then(data => {
          const now = Date.now();
          dispatch({
            type: responseType,
            payload: data,
            meta: {
              schema,
              url,
              date: now,
              expiresAt: now + dataExpiryLength,
            },
          });
          return data;
        })
        .catch(error => {
          const now = Date.now();
          dispatch({
            type: responseType,
            payload: error,
            meta: {
              schema,
              url,
              date: now,
              expiresAt: now + errorExpiryLength,
            },
            error: true,
          });
          throw error;
        });
    let promise;
    if (throttle) {
      promise = this.throttle(url, deferedFetch);
    } else {
      promise = deferedFetch();
    }
    promise.then(data => resolve(data)).catch(error => reject(error));
    return promise;
  }

  /** Called when middleware intercepts a receive action.
   *
   * Will resolve the promise associated with receive url.
   */
  protected handleReceive(action: ReceiveAction) {
    const completePromise = () => {
      // this can still turn out to be untrue since this is async
      if (action.meta.url in this.fetched) {
        let promiseHandler: (value?: any) => void;
        if (action.error) {
          promiseHandler = this.rejectors[action.meta.url];
        } else {
          promiseHandler = this.resolvers[action.meta.url];
        }
        promiseHandler(action.payload);
        // since we're resolved we no longer need to keep track of this promise
        this.clear(action.meta.url);
      }
    };
    // TODO: this should call after the reducer has been updated
    // in all concurrent fibers
    RIC(completePromise, { timeout: 1000 });
  }

  /** Attaches NetworkManager to store
   *
   * Intercepts 'rest-hooks/fetch' actions to start requests.
   *
   * Resolve/rejects a request when matching 'rest-hooks/receive' event
   * is seen.
   */
  getMiddleware = memoize(function<T extends NetworkManager>(this: T) {
    return <R extends React.Reducer<any, any>>({
      dispatch,
    }: MiddlewareAPI<R>) => {
      return (next: React.Dispatch<React.ReducerAction<R>>) => (
        action: React.ReducerAction<R>,
      ) => {
        switch (action.type) {
        case 'rest-hooks/fetch':
          this.handleFetch(action, dispatch);
          return;
        case 'rest-hooks/purge':
        case 'rest-hooks/rpc':
        case 'rest-hooks/receive':
          // only receive after new state is computed
          next(action);
          if (action.meta.url in this.fetched) {
            this.handleReceive(action);
          }
          return;
        default:
          next(action);
          return;
        }
      };
    };
  });

  /** Ensures only one request for a given url is in flight at any time
   *
   * Uses url as key to either retrieve in-flight promise, or if not
   * create a new promise and call fetch.
   *
   * Note: The new promise is not actually tied to fetch at all,
   * but is resolved when the expected 'recieve' action is processed.
   * This ensures promises are resolved only once their data is processed
   * by the reducer.
   */
  protected throttle(url: string, fetch: () => Promise<any>) {
    // we're already fetching so reuse the promise
    if (url in this.fetched) {
      return this.fetched[url];
    }

    this.fetched[url] = new Promise((resolve, reject) => {
      this.resolvers[url] = resolve;
      this.rejectors[url] = reject;
    });

    // since our real promise is resolved via the wrapReducer(),
    // we should just stop all errors here.
    // TODO: decouple this from useFetcher() (that's what's dispatching the error the resolves in here)
    RIC(() => fetch().catch(() => null), { timeout: 500 });

    return this.fetched[url];
  }
}
