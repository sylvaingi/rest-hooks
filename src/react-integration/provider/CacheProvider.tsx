import React, { ReactNode, useEffect } from 'react';
import { StateContext, DispatchContext } from '../context';
import masterReducer, {
  initialState as defaultState,
} from '../../state/reducer';
import NetworkManager from '../../state/NetworkManager';
import SubscriptionManager from '../../state/SubscriptionManager';
import PollingSubscription from '../../state/PollingSubscription';
import createEnhancedReducerHook from './middleware';
import { State } from '../../types';

interface ProviderProps {
  children: ReactNode;
  manager?: NetworkManager;
  subscriptionManager?: SubscriptionManager<any>;
  initialState?: State<unknown>;
}
/** Controller managing state of the REST cache and coordinating network requests. */
export default function CacheProvider({
  children,
  manager = new NetworkManager(),
  subscriptionManager = new SubscriptionManager(PollingSubscription),
  initialState = defaultState,
}: ProviderProps) {
  // TODO: option to use redux
  const useEnhancedReducer = createEnhancedReducerHook(
    manager.getMiddleware(),
    subscriptionManager.getMiddleware(),
  );
  const [state, dispatch] = useEnhancedReducer(masterReducer, initialState);

  // if we change out the manager we need to make sure it has no hanging async
  useEffect(() => {
    return () => {
      manager.cleanup();
    };
  }, [manager]);
  useEffect(() => {
    return () => {
      subscriptionManager.cleanup();
    };
  }, [subscriptionManager]);

  return (
    <DispatchContext.Provider value={dispatch}>
      <StateContext.Provider value={state}>{children}</StateContext.Provider>
    </DispatchContext.Provider>
  );
}
