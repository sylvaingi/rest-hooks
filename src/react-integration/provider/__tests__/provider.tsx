import React, { useContext } from 'react';
import { cleanup, act, render } from 'react-testing-library';

import { DispatchContext, StateContext } from '../../context';
import { CoolerArticleResource } from '../../../__tests__/common';
import CacheProvider from '../CacheProvider';
import NetworkManager from '../../../state/NetworkManager';

afterEach(cleanup);

describe('<CacheProvider />', () => {
  it('should not change dispatch function on re-render', () => {
    let dispatch;
    let count = 0;
    function DispatchTester() {
      dispatch = useContext(DispatchContext);
      count++;
      return null;
    }
    const chil = <DispatchTester />;
    const tree = <CacheProvider>{chil}</CacheProvider>;
    const { rerender } = render(tree);
    expect(dispatch).toBeDefined();
    const curDisp = dispatch;
    rerender(tree);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    rerender(<CacheProvider>{chil}</CacheProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    const manager = new NetworkManager();
    rerender(<CacheProvider manager={manager}>{chil}</CacheProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    rerender(
      <DispatchContext.Provider value={() => null}>
        {chil}
      </DispatchContext.Provider>,
    );
    expect(curDisp).not.toBe(dispatch);
    expect(count).toBe(2);
  });
  it('should change state', () => {
    let dispatch: any, state;
    let count = 0;
    function ContextTester() {
      dispatch = useContext(DispatchContext);
      state = useContext(StateContext);
      count++;
      return null;
    }
    const chil = <ContextTester />;
    const tree = <CacheProvider>{chil}</CacheProvider>;
    render(tree);
    expect(dispatch).toBeDefined();
    expect(state).toBeDefined();
    const action = {
      type: 'rest-hooks/receive',
      payload: { id: 5, title: 'hi', content: 'more things here' },
      meta: {
        schema: CoolerArticleResource.getEntitySchema(),
        url: CoolerArticleResource.url({ id: 5 }),
        mutate: false,
        date: 50,
        expiresAt: 55,
      },
    };
    act(() => {
      dispatch(action);
    });
    expect(count).toBe(2);
    expect(state).toMatchSnapshot();
  });
});
