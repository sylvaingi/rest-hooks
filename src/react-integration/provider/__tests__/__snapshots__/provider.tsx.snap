// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CacheProvider /> should change state 1`] = `
Object {
  "entities": Object {
    "http://test.com/article-cooler/": Object {
      "5": CoolerArticleResource {
        "author": null,
        "content": "more things here",
        "id": 5,
        "tags": Array [],
        "title": "hi",
      },
    },
  },
  "meta": Object {
    "http://test.com/article-cooler/5": Object {
      "date": 50,
      "expiresAt": 55,
    },
  },
  "results": Object {
    "http://test.com/article-cooler/5": "5",
  },
}
`;
